src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -1002,24 +1004,44 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
         /// <param name="input">The input string.</param>
         /// <param name="regex">A Regex instance.</param>
         /// <param name="substitute">The substitute value.</param>
+        /// <param name="cachedReplacementString">Cached replacement string.</param>
+        /// <param name="cachedMatchEvaluator">Cached MatchEvaluator.</param>
         /// <returns>The result of the regex.Replace operation.</returns>
-        private static object ReplaceOperatorImpl(ExecutionContext context, string input, Regex regex, object substitute)
+        private static object ReplaceOperatorImpl(
+            ExecutionContext context,
+            string input,
+            Regex regex,
+            object substitute,
+            ref string cachedReplacementString,
+            ref MatchEvaluator cachedMatchEvaluator)
         {
+            // After first call we will have necessarily either cachedReplacementString or cachedMatchEvaluator.
+            if (cachedReplacementString is not null)
+            {
+                return regex.Replace(input, cachedReplacementString);
+            }
+
+            if (cachedMatchEvaluator is not null)
+            {
+                return regex.Replace(input, cachedMatchEvaluator);
+            }
```

Comment:
I'm not sure I understand the semantics here.

Let's say we have an `IEnumerable` of two strings. On the first string we hit line 1031, and cache that string `replacement`.

On the second string, it looks like we have a cached value from the first call, so we ignore the second string and immediately use the first again?

How does that work?

Reply:
Replacement string is not _input_ string - I guess you lost this.

Reply:
Ah I see. Ok, so why on earth does the implementation do it like this? It's pretty easy to convert the substitute to a know result ahead of time without complicated caching logic. Once again, we've very seriously forgotten how to use object-oriented programming.

I think we need to:

- Create the substitute object once, above the loop
- Then pass it into the `ReplaceOperatorImpl` method

That's much much simpler than anything else.

```csharp
private abstract class RegexReplacer
{
    protected RegexReplacer(Regex regex)
    {
        ReplacementRegex = regex;
    }

    public abstract string Replace(string input);

    protected Regex ReplacementRegex { get; }
}

private class StringReplacer : RegexReplacer
{
    private readonly string _replacementString;

    public StringReplacer(Regex regex, string replacement)
        : base(regex)
    {
        _replacementString = str;
    }

    public override string Replace(string input)
    {
        return ReplacementRegex.Replace(input, _replacementString);
    }
}

private class MatchEvaluatorReplacer : RegexReplacer
{
    private readonly MatchEvaluator _replacementMatchEvaluator;

    public MatchEvaluatorReplacer(Regex regex, MatchEvaluator replacement)
        : base(regex)
    {
        _replacementMatchEvaluator = replacement;
    }

    public override string Replace(string input)
    {
        return ReplacementRegex.Replace(input, _replacementMatchEvaluator);
    }
}

...

private static RegexReplacer GetSubstituteReplacer(
    ExecutionContext context,
    Regex regex,
    object substitute)
{
    switch (substitute)
    {
        // When the substitution is a string, we can skip all the conversion logic
        case string replacement:
            return new StringReplacer(regex, replacement);

        // ScriptBlocks implement custom regex evaluators, so we must handle them specially
        case ScriptBlock sb:
            return new MatchEvaluatorReplacer(regex, GetMatchEvaluator(context, sb));
    }

    // Try once more to convert the substitute to a match evaluator
    if (LanguagePrimitives.TryConvertTo(subtitute, out MatchEvaluator matchEvaluator))
    {
        return new MatchEvaluatorReplacer(regex, matchEvaluator);
    }

    // Finally, we just treat the substitute as a string
    return new StringReplacer(regex, PSObject.ToStringParser(context, substitute));
}

...

// <main method where we're performing the logic>
{
    RegexReplacer replacer = GetSubstituteReplacer(context, rr, substitute);
    if (list == null)
    {
        ...
        return replacer.Replace(lvalString);
    }
    else
    {
        List<object> resultList = new List<object>();
        while (ParserOps.MoveNext(context, errorPosition, list))
        {
            string lvalString = PSObject.ToStringParser(context, ParserOps.Current(errorPosition, list));
            resultList.Add(replacer.Replace(lvalString);
        }
    }
}
```

Reply:
I think creating classes is redundant in the context so I combined the proposal with your previous based on struct.

Reply:
But it's not redundant. The class encapsulates a pre-computed result. We shouldn't be revisiting that `switch` statement inside the loop.

Reply:
Ok I can live with your most recent proposal